components/eam/src/physics/cam/hetfrz_classnuc.F90

```diff
@@ -310,272 +250,467 @@ subroutine hetfrz_classnuc_calc( &
+   sum_imm_dust_a1 = 0._r8
+   sum_imm_dust_a3 = 0._r8
+   do ibin = itheta_bin_beg, itheta_bin_end-1
+      sum_imm_dust_a1 = sum_imm_dust_a1 + 0.5_r8*((pdf_imm_theta(ibin)*exp(-dim_Jimm_dust_a1(ibin)*deltat)+ &
+                        pdf_imm_theta(ibin+1)*exp(-dim_Jimm_dust_a1(ibin+1)*deltat)))*pdf_d_theta
+      sum_imm_dust_a3 = sum_imm_dust_a3 + 0.5_r8*((pdf_imm_theta(ibin)*exp(-dim_Jimm_dust_a3(ibin)*deltat)+ &
+                        pdf_imm_theta(ibin+1)*exp(-dim_Jimm_dust_a3(ibin+1)*deltat)))*pdf_d_theta
+   enddo
+
+   
+   if (sum_imm_dust_a1 > 0.99_r8) then
+      sum_imm_dust_a1 = 1.0_r8
+   endif
+   if (sum_imm_dust_a3 > 0.99_r8) then
+      sum_imm_dust_a3 = 1.0_r8
+   endif
+ 
+
+   if (do_bc) frzbcimm = frzbcimm+MIN(limfacbc*total_cloudborne_aer_num(id_bc)/deltat, &
+                                     total_cloudborne_aer_num(id_bc)/deltat*(1._r8-exp(-Jimm_bc*deltat)))
+
+   if (do_dst1) frzduimm = frzduimm+MIN(1*total_cloudborne_aer_num(id_dst1)/deltat,        &
                                  total_cloudborne_aer_num(id_dst1)/deltat*(1._r8-sum_imm_dust_a1))
-         if (do_dst3) frzduimm = frzduimm+MIN(1*total_cloudborne_aer_num(id_dst3)/deltat,        &
+   if (do_dst3) frzduimm = frzduimm+MIN(1*total_cloudborne_aer_num(id_dst3)/deltat,        &
                                  total_cloudborne_aer_num(id_dst3)/deltat*(1._r8-sum_imm_dust_a3))
-      end if
-
-   else
-      if (do_bc) frzbcimm = frzbcimm+MIN(limfacbc*fn(id_bc)*total_aer_num(id_bc)/deltat, & 
-                            fn(id_bc)*total_aer_num(id_bc)/deltat*(1._r8-exp(-Jimm_bc*deltat))) 
-
-      if (.not. pdf_imm_in) then
-         if (do_dst1) frzduimm = frzduimm+MIN(1*fn(id_dst1)*total_aer_num(id_dst1)/deltat, &
-                                 fn(id_dst1)*total_aer_num(id_dst1)/deltat*(1._r8-exp(-Jimm_dust_a1*deltat)))
-         if (do_dst3) frzduimm = frzduimm+MIN(1*fn(id_dst3)*total_aer_num(id_dst3)/deltat, &
-                                 fn(id_dst3)*total_aer_num(id_dst3)/deltat*(1._r8-exp(-Jimm_dust_a3*deltat)))
-      else
-         if (do_dst1) frzduimm = frzduimm+MIN(1*fn(id_dst1)*total_aer_num(id_dst1)/deltat,        &
-                                 fn(id_dst1)*total_aer_num(id_dst1)/deltat*(1._r8-sum_imm_dust_a1))
-         if (do_dst3) frzduimm = frzduimm+MIN(1*fn(id_dst3)*total_aer_num(id_dst3)/deltat,        &
-                                 fn(id_dst3)*total_aer_num(id_dst3)/deltat*(1._r8-sum_imm_dust_a3))
-      end if
-   end if
-
-   if (t > 263.15_r8) then
+
+
+   if (temperature > 263.15_r8) then
       frzduimm = 0._r8
       frzbcimm = 0._r8
-   end if
-
-   !!! !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
-   !   Deposition nucleation
-   !!! !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
-   ! critical germ size
-   ! assume 98% RH in mixed-phase clouds (Korolev & Isaac, JAS 2006)
-   rgdep=2*vwice*sigma_iv/(kboltz*t*LOG(rhwincloud*supersatice)) 
+   endif
+
+end subroutine calculate_hetfrz_immersion_nucleation
```

Comment:
Subroutine `calculate_hetfrz_immersion_nucleation` has repeated calculations. You can create small functions for those or leave them as is. Repeated code is prone to bugs and hard to maintain.

Reply:
added functions get_dg0imm and get_Aimm.